clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -334,11 +334,46 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
     llvm_unreachable("NYI");
   }
   mlir::Value VisitUnaryPlus(const UnaryOperator *E) {
-    llvm_unreachable("NYI");
+    // NOTE(cir): QualType function parameter still not used, so don´t replicate
+    // it here yet.
+    QualType promotionTy = getPromotionType(E->getSubExpr()->getType());
+    auto result = VisitPlus(E, promotionTy);
+    if (result && !promotionTy.isNull())
+      assert(0 && "not implemented yet");
+
+    // NOTE(cir): we build the unary operation here, further lowering
+    // is performed by the LowerToLLVM passes.
+    return buildUnaryOp(E, mlir::cir::UnaryOpKind::Plus, result);
+  }
+
+  mlir::Value VisitPlus(const UnaryOperator *E, QualType PromotionType) {
+    // This differs from gcc, though, most likely due to a bug in gcc.
+    TestAndClearIgnoreResultAssign();
+    if (!PromotionType.isNull())
+      assert(0 && "scalar promotion not implemented yet");
+    return Visit(E->getSubExpr());
   }
+
   mlir::Value VisitUnaryMinus(const UnaryOperator *E) {
-    llvm_unreachable("NYI");
+    // NOTE(cir): QualType function parameter still not used, so don´t replicate
+    // it here yet.
+    QualType promotionTy = getPromotionType(E->getSubExpr()->getType());
+    auto result = VisitMinus(E, promotionTy);
+    if (result && !promotionTy.isNull())
+      assert(0 && "not implemented yet");
+
+    // NOTE(cir): we build the unary operation here, further lowering
+    // is performed by the LowerToLLVM passes.
+    return buildUnaryOp(E, mlir::cir::UnaryOpKind::Minus, result);
   }
+
+  mlir::Value VisitMinus(const UnaryOperator *E, QualType PromotionType) {
+    TestAndClearIgnoreResultAssign();
+    if (!PromotionType.isNull())
+      assert(0 && "scalar promotion not implemented yet");
+    return Visit(E->getSubExpr());
```

Comment:
Might also be worth a short comment explaining that during LLVM codegen float uses fneg and a emit sub is used.

Reply:
Added

```
    // NOTE: LLVM codegen will lower this directly to either a FNeg
    // or a Sub instruction.  In CIR this will be handled later in LowerToLLVM.
```

but removed the comments I had in VisitUnaryPlus and VisitUnaryMinus.